File: 143.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if (!head->next)
            return;
        deque<ListNode *> dq;
        ListNode *cur = head;
        while (cur != nullptr) {
            dq.push_back(cur);
            cur = cur -> next;
        }
        
        int count = 0;
        ListNode *next;
        cur = dq.front();
        dq.pop_front();
        while (!dq.empty()) {
            if (count % 2 == 0) {
                next = dq.back();
                dq.pop_back();
            } else {
                next = dq.front();
                dq.pop_front();
            }
            cur->next = next;
            cur = cur->next;
            cur->next = nullptr;
            count++;
        }
        
    }
};

File: go/143.go
package main

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func reorderList(head *ListNode) {
	if head == nil || head.Next == nil {
		return
	}

	var cur *ListNode = head.Next
	var st []*ListNode = make([]*ListNode, 0)

	for cur != nil {
		st = append(st, cur)
		cur = cur.Next
	}

	cur = head
	for len(st) > 1 {
		cur.Next = st[len(st)-1]
		cur = cur.Next
		cur.Next = st[0]
		st = st[1 : len(st)-1]
		cur = cur.Next
	}
	if len(st) == 1 {
		cur.Next = st[0]
		cur = cur.Next
	}
	cur.Next = nil

}
